File: 4.8-bc/py/stage-3-hexagonal/app/billing/domain/subscription.py
"""Billing domain — Subscription entity + SubscriptionCreated event."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from uuid import UUID

from app.shared.events import Event


@dataclass
class Subscription:
    id: UUID
    user_id: UUID
    plan: str
    next_charge_at: datetime
    created_at: datetime


class InvalidPlanError(Exception):
    """billing.invalid_plan"""


@dataclass
class SubscriptionCreated(Event):
    """Domain event — Notifications subscribes for the welcome letter."""

    subscription_id: UUID
    user_id: UUID
    plan: str
    at: datetime

    @classmethod
    def name(cls) -> str:
        return "billing.SubscriptionCreated"

File: 4.8-bc/ts/stage-3-hexagonal/src/billing/domain/subscription.ts
// Billing Bounded Context, доменні типи й event.
import type { DomainEvent } from "../../shared/events/bus.js";

export interface Subscription {
  id: string;
  userId: string;
  plan: string;
  nextChargeAt: Date;
  createdAt: Date;
}

// SubscriptionCreated — Notifications підписується для welcome-letter.
export class SubscriptionCreated implements DomainEvent {
  static readonly EVENT_NAME = "billing.SubscriptionCreated";
  readonly name: string = SubscriptionCreated.EVENT_NAME;

  constructor(
    public readonly subscriptionId: string,
    public readonly userId: string,
    public readonly plan: string,
    public readonly at: Date,
  ) {}
}
